tests: net: Increase network buffer counts for Atmel SAM-E70

jukkar opened the pull request:
In some hw like Atmel SAM-E70, the original amount of network
buffers is too small so increasing it slightly.
Also generate proper error if this happens in the test so it
is easier to figure out.

Fixes #8963

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

nashif commented on tests/net/trickle/prj.conf:
```diff
@@ -10,10 +10,10 @@ CONFIG_ENTROPY_GENERATOR=y
+CONFIG_NET_PKT_TX_COUNT=10
+CONFIG_NET_PKT_RX_COUNT=10
+CONFIG_NET_BUF_RX_COUNT=20
+CONFIG_NET_BUF_TX_COUNT=20
```
Can we just use safe defaults and stop setting those in project configs? Looks like we keep hitting this and now we are fixing it until we hit it again next time...

pfalcon replied:
We do have safe defaults now, we just have "broken" (in a sense) drivers which can't work with conservative defaults and want bloat, bloat, bloat. And this is a test anyway, so may have special needs.

nashif replied:
safe means it should also work with "broken" drivers. And given this is a test, we should apply those values and if someone wants to optimize, they can tune those down. safe default also means most people should not have to worry about those values, they should just take a the default.

pfalcon replied:
For me, "safe" means "the minimal settings with which the IP stack can work with, and helping to detect broken drivers and stuff".

pfalcon replied:
Masking the issues in the drivers and stack by throwing more memory will never give us a working system. Actually, it already gives the opposite, see e.g. https://github.com/zephyrproject-rtos/zephyr/pull/7831

nashif replied:
> For me, "safe" means "the minimal settings with which the IP stack can work with, and helping to detect broken drivers and stuff".

For me it means the minimal settings that covers maximal HW and SW combinations, so users and developers are not left guessing what values to use.

Maybe:
add 3 options (NET_PRODUCTION,  NET_DEVELOPMENT,  NET_TESTING )and select values based on those options, for example:
```
config NET_BUF_RX_COUNT
        int "How many network buffers are allocated for receiving data"
        default 16 if NET_PRODUCTION
        default 20 if NET_DEVELOPMENT
        default 36 if NET_TESTING
        default 36 if NET_L2_ETHERNET
        help
          Each data buffer will occupy CONFIG_NET_BUF_DATA_SIZE + smallish
          header (sizeof(struct net_buf)) amount of data.
```

nashif replied:
my request really has nothing to do with the atmel, from my observing point of view there is no real science in how those values are being set, sometimes they are being copied over from one sample to the next, increased/decreased during debugging and left as is and basically completely random at some times. 

```
git grep -h  NET_BUF_RX_COUNT tests/ samples/ | sort | wc -l
101
```
and
```
git grep -h  NET_BUF_RX_COUNT tests/ samples/ | sort | uniq
CONFIG_NET_BUF_RX_COUNT=10
CONFIG_NET_BUF_RX_COUNT=100
CONFIG_NET_BUF_RX_COUNT=128
CONFIG_NET_BUF_RX_COUNT=15
CONFIG_NET_BUF_RX_COUNT=16
CONFIG_NET_BUF_RX_COUNT=20
CONFIG_NET_BUF_RX_COUNT=200
CONFIG_NET_BUF_RX_COUNT=25
CONFIG_NET_BUF_RX_COUNT=28
CONFIG_NET_BUF_RX_COUNT=3
CONFIG_NET_BUF_RX_COUNT=30
CONFIG_NET_BUF_RX_COUNT=32
CONFIG_NET_BUF_RX_COUNT=34
CONFIG_NET_BUF_RX_COUNT=35
CONFIG_NET_BUF_RX_COUNT=36
CONFIG_NET_BUF_RX_COUNT=40
CONFIG_NET_BUF_RX_COUNT=48
CONFIG_NET_BUF_RX_COUNT=5
CONFIG_NET_BUF_RX_COUNT=50
CONFIG_NET_BUF_RX_COUNT=6
CONFIG_NET_BUF_RX_COUNT=60
CONFIG_NET_BUF_RX_COUNT=64
CONFIG_NET_BUF_RX_COUNT=7
CONFIG_NET_BUF_RX_COUNT=720
CONFIG_NET_BUF_RX_COUNT=80
CONFIG_NET_BUF_RX_COUNT=96
```

Beside the kconfig documentation there is not much being told how fine tune those and what guidelines are there to follow when setting them, so provide "safe" values that works for the 80% and additional ultra safe for testing and development for the remaining 20% will help. Documenting this and setting some guidelines would help even more.


Disclaimer: maybe all the values above are not random after all and someone spent lots of time calculating and testing, so take no offence please :)

jukkar replied:
> Disclaimer: maybe all the values above are not random after all and someone spent lots of time calculating and testing

The values seem random, but at least I have tried to think about the use cases in various samples when deciding the values. It is quite hard actually to create values for PRODUCTION, DEVELOPMENT or TESTING scenarios.
Example values:
CONFIG_NET_PKT_RX_COUNT=1
CONFIG_NET_BUF_RX_COUNT=2

This means that we are able to receive one network packet, and max size of that packet is by default 2 * 128 bytes. So the values really depend on protocol used, usage scenario etc. Also network stack needs to do various things on its own like IPv6 neighbor discovery or IPv4 ARP, we need a bit more buffers in practice to make things work properly. But what this "more" bufs means is the question here.

For unit tests we typically do not do IPv6 ND etc. which is why the limits are quite small in prj.conf files under tests/net. Unfortunately some of these tests are now run in real hw which seems quite unnecessary IMHO and complicates things.

jukkar commented:
Actually I do not see much point running these tests in real hw. The values I changed here were originally optimized for qemu / native_posix use. As these are test programs it should not matter much what values we have here.

nashif commented:
> Actually I do not see much point running these tests in real hw. The values I changed here were originally optimized for qemu / native_posix use. As these are test programs it should not matter much what values we have here.

Then please make them qemu only, I can live with that.

jukkar committed:
tests: net: Print proper error if we run out of net bufs

Generate proper error if we run out of network buffers which means
so that the test can fail properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: Run various tests only in qemu_x86 or native_posix

No need to run the tests in real hardware as there we can run
out of network buffers that happened with Atmel SAM-E70.

Fixes #8966

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated the tests to run only in qemu_x87 or native_posix.
